modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -236,13 +242,54 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
         }
     }
 
+    private void doWriteChunkedContinuation(ChannelHandlerContext ctx, Netty4ChunkedHttpContinuation continuation, ChannelPromise promise)
+        throws IOException {
+        final PromiseCombiner combiner = continuation.combiner();
+        assert currentChunkedWrite == null;
+        final var responseBody = continuation.body();
+        currentChunkedWrite = new ChunkedWrite(combiner, promise, responseBody);
+        // NB "writable" means there's space in the downstream ChannelOutboundBuffer, we aren't trying to saturate the physical channel.
+        while (ctx.channel().isWritable()) {
+            if (writeChunk(ctx, combiner, responseBody)) {
+                finishChunkedWrite();
+                return;
+            }
+        }
+    }
+
     private void finishChunkedWrite() {
         assert currentChunkedWrite != null;
         assert currentChunkedWrite.responseBody().isDone();
         final var finishingWrite = currentChunkedWrite;
+        final var endOfResponse = finishingWrite.responseBody().isEndOfResponse();
         currentChunkedWrite = null;
-        writeSequence++;
-        finishingWrite.combiner.finish(finishingWrite.onDone());
+        if (endOfResponse) {
+            writeSequence++;
+            finishingWrite.combiner.finish(finishingWrite.onDone());
+        } else {
+            final var channel = finishingWrite.onDone().channel();
+            ActionListener.run(ActionListener.assertOnce(new ActionListener<>() {
+                @Override
+                public void onResponse(ChunkedRestResponseBody continuation) {
+                    channel.writeAndFlush(
+                        new Netty4ChunkedHttpContinuation(writeSequence, continuation, finishingWrite.combiner()),
+                        finishingWrite.onDone() // pass the terminal listener/promise along the line
+                    );
```

Comment:
I don't think so, this code runs on the event loop already so these writes will still either complete or fail properly even if we're in the process of shutting down.

Reply:
Ah yes you're quite right, thanks

Reply:
I've applied a fix in 6fac79de44bf4370837c76442b48e2b404beb5f6 but it had to be a little different from #105306 since we don't have a listener to wrap any more. But then I realised we could do the same thing at the outer layer too, and extract a utility, and in fact what we do today at the outer layer is also a little questionable too. I made the relevant changes in https://github.com/elastic/elasticsearch/pull/105486 and will migrate to that utility once it's available here.